src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,263 @@
+use Symfony\Component\AutoMapper\Extractor\FromTargetMappingExtractor;
+use Symfony\Component\AutoMapper\Extractor\SourceTargetMappingExtractor;
+use Symfony\Component\AutoMapper\Generator\Generator;
+use Symfony\Component\AutoMapper\Loader\ClassLoaderInterface;
+use Symfony\Component\AutoMapper\Loader\EvalLoader;
+use Symfony\Component\AutoMapper\Transformer\ArrayTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\BuiltinTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\ChainTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\DateTimeTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\MultipleTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\NullableTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\ObjectTransformerFactory;
+use Symfony\Component\AutoMapper\Transformer\UniqueTypeTransformerFactory;
+use Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor;
+use Symfony\Component\PropertyInfo\Extractor\ReflectionExtractor;
+use Symfony\Component\PropertyInfo\PropertyInfoExtractor;
+use Symfony\Component\Serializer\Mapping\ClassDiscriminatorFromClassMetadata;
+use Symfony\Component\Serializer\Mapping\Factory\ClassMetadataFactory;
+use Symfony\Component\Serializer\Mapping\Loader\AnnotationLoader;
+use Symfony\Component\Serializer\NameConverter\AdvancedNameConverterInterface;
+
+/**
+ * Maps a source data structure (object or array) to a target one.
+ *
+ * @expiremental in 5.1
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+final class AutoMapper implements AutoMapperInterface, AutoMapperRegistryInterface, MapperGeneratorMetadataRegistryInterface
+{
+    /** @var MapperGeneratorMetadataInterface[] */
+    private $metadata = [];
+
+    /** @var GeneratedMapper[] */
+    private $mapperRegistry = [];
+
+    private $classLoader;
+
+    private $mapperConfigurationFactory;
+
+    public function __construct(ClassLoaderInterface $classLoader, MapperGeneratorMetadataFactoryInterface $mapperConfigurationFactory = null)
+    {
+        $this->classLoader = $classLoader;
+        $this->mapperConfigurationFactory = $mapperConfigurationFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register(MapperGeneratorMetadataInterface $metadata): void
+    {
+        $this->metadata[$metadata->getSource()][$metadata->getTarget()] = $metadata;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMapper(string $source, string $target): MapperInterface
+    {
+        $metadata = $this->getMetadata($source, $target);
+
+        if (null === $metadata) {
+            throw new NoMappingFoundException('No mapping found for source '.$source.' and target '.$target);
+        }
+
+        $className = $metadata->getMapperClassName();
+
+        if (\array_key_exists($className, $this->mapperRegistry)) {
+            return $this->mapperRegistry[$className];
+        }
+
+        if (!class_exists($className)) {
```

Comment:
Shouldn't you write `\class_exists()` ?

Reply:
Hum it should not in fact https://fabbot.io/report/symfony/symfony/30248/ff422e426c80bd132b64c4ad6965700b719469e4, think it's not actually optimized by php or maybe another reason.

Reply:
I can't open the link. I usually make every function call explicit, no matter if it's optimised or not. For a long time, `\array_key_exists()` was not optimised, but now, it is an opcode which yields much better performance: people that made the `\` explicit have a free performance benefit, other don't. I like to write future-proof code in most case it doesn't hurt, and sometime, it yields unexpected benefits.

But if benchmarks says otherwise, I'm very curious to know why!